docs/architecture/pk-4.md

```diff
@@ -0,0 +1,352 @@
+
+### Key parsing and writing
+
+Keep:
+```
+mbedtls_pk_parse_key()
+mbedtls_pk_parse_public_key()
+mbedtls_pk_write_key_der()
+mbedtls_pk_write_pubkey_der()
+mbedtls_pk_write_pubkey_pem()
+mbedtls_pk_write_key_pem()
+```
+
+Keep:
+```
+mbedtls_pk_parse_keyfile()
+mbedtls_pk_parse_public_keyfile()
+```
+
+### Signature conveniences
+
+#### Basic signature functionality
+
+Keep the following:
+```
+MBEDTLS_PK_SIGNATURE_MAX_SIZE
+mbedtls_pk_verify()
+mbedtls_pk_sign()
+```
+
+ACTION: remove `MBEDTLS_ERR_PK_SIG_LEN_MISMATCH`.
+
+#### Restartable signature functionality
+
+It's more convenient to keep using PK for restartable signature in X.509, for the same reason as non-restartable signature. So restartable PK should keep existing, so we might as well keep it public.
+
+Keep:
+```
+mbedtls_pk_restart_ctx
+mbedtls_pk_restart_init()
+mbedtls_pk_restart_free()
+mbedtls_pk_verify_restartable()
+mbedtls_pk_sign_restartable()
+```
+
+### Removals
+
+#### Privatization
+
+ACTION: Move all private API elements to an private header:
+
+```
+mbedtls_pk_type_t
+mbedtls_pk_rsassa_pss_options
+mbedtls_pk_debug_type
+mbedtls_pk_debug_item
+MBEDTLS_PK_DEBUG_MAX_ITEMS
+mbedtls_pk_info_from_type()
+mbedtls_pk_setup()
+mbedtls_pk_get_len()
+mbedtls_pk_can_do()
+mbedtls_pk_verify_ext()
+mbedtls_pk_sign_ext()
+mbedtls_pk_decrypt()
+mbedtls_pk_encrypt()
+mbedtls_pk_debug()
+mbedtls_pk_get_name()
+mbedtls_pk_get_type()
+mbedtls_pk_rsa()
+mbedtls_pk_ec()
+mbedtls_pk_parse_subpubkey()
+mbedtls_pk_write_pubkey()
+```
+
+Follow-up: [Make private API elements internal](#make-private-API-elements-internal)
+
+#### Documentation update after privatization
+
+ACTION: remove all mentions of private API elements from the public documentation.
+
+This includes both direct mentions (where a type name, constant name or function name is mentioned), and indirect mentions (e.g. “verify\_ext”, “context has been set up”).
+
+ACTION: update unit tests to validate the new interfaces.
+
+#### Remove RSA-ALT
+
+ACTION: Remove the option `MBEDTLS_PK_RSA_ALT_SUPPORT` and all code guarded by it, as well as `MBEDTLS_PK_RSA_ALT`.
+
+### Documentation updates
+
+ACTION: update the PSA transition guide.
+
+ACTION: write changelog entries.
+
+## Open questions
+
+### Parsing and writing
+
+#### Remove file parsing functions?
+
+Should we remove the file parsing functions `mbedtls_pk_parse_keyfile()` and `mbedtls_pk_parse_public_keyfile()`? They look misplaced in a library that generally doesn't access files. But it isn't really difficult to keep them.
+
+These functions are used in test code and sample programs.
+
+### Resource management
+
+#### Access the underlying PSA key of a PK context
+
+Should we provide a function to access the underlying PSA key of a PSA context, if there is one?
```

Comment:
I'm inclined to say no initially. We can add functions at any time if we feel a need in the future.